toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,274 @@
+
+Security-wise it is easier for an attacker to probe the environment by simply doing a port scan and then testing direct attacks on the SSH proxy:
+
+```
+nmap api.cf.eu10.hana.ondemand.com
+
+Starting Nmap 7.94 ( https://nmap.org ) at 2024-04-08 10:29 CEST
+Nmap scan report for api.cf.eu10.hana.ondemand.com (3.124.222.77)
+Host is up (0.031s latency).
+Other addresses for api.cf.eu10.hana.ondemand.com (not scanned): 3.122.209.241 3.124.208.223
+rDNS record for 3.124.222.77: ec2-3-124-222-77.eu-central-1.compute.amazonaws.com
+Not shown: 997 filtered tcp ports (no-response)
+PORT     STATE SERVICE
+80/tcp   open  http
+443/tcp  open  https
+2222/tcp open  EtherNetIP-1
+```
+
+then proceed with connecting to the port 2222:
+
+```
+nc api.cf.eu10.hana.ondemand.com 2222
+
+SSH-2.0-diego-ssh-proxy
+```
+
+the SSH proxy greets with the SSH-2.0 version, allowing the attacker to try all kinds of SSH-2.0 related probes.
+
+Another facet of the problem is that many customers require explicit allow-listing of external web sites from their on-prem environments.
+Port 443 is usually not a problem as many firewalls open it by default, but port 2222 likely is frowned upon and cause for concern.
+
+## Proposal
+
+### Note on Security
+The proposal does not aim to solve security issues of (CF) SSH per se. Tunneling a protocol through another does not make it inherently more secure. Any weakness of SSH within the CF SSH proxy will still be present even if the RFC is accepted. The main goal is to reduce the number of open ports required. Although, there is one minor security improvement as the SSH traffic will then be also encrypted by TLS within the WebSocket tunnel.
+
+
+### Architecture Changes
+
+![architecture-changes](rfc-draft-ssh-over-ws/cf-ssh-over-websocket.drawio.png)
+
+There are three scenarios:
+
+#### Scenario A: Legacy SSH
+This is the default scenario as of today. All changed components MUST ensure this scenario still works. It SHOULD be tested with legacy CLI versions such as v6 which are no longer updated. This scenario includes regular SSH clients like OpenSSH.
+
+#### Scenario B: Dual SSH / WS
+This scenario MAY be enabled by operators. It allows legacy SSH as well as pure WebSocket-based SSH in one environment. The CF CLI MUST understand the legacy flow from scenario A. The CF CLI SHOULD understand the new flow using a WebSocket tunnel. Other components MUST support both scenarios.
+
+#### Scenario C: WS only
+This scenario MAY be enabled by operators. It turns of the legacy scenario A and allows operators to close the public port 2222 on the load balancer. The CF CLI MUST understand the new flow using a WebSocket tunnel. Other components MUST support the new flow and MAY support both scenarios.
+
+
+### Changes per Component (Scenario B and C)
+
+#### CF CLI
+- The CLI MUST query the "v2/info" endpoint of Cloud Controller.
```

Comment:
CLI v7 and v8 use the `/` endpoint of CC to learn about the foundation.
The ssh info looks like this:
```
{
  "links": {
    "app_ssh": {
      "href": "ssh.cf-app.com:2222",
      "meta": {
        "host_key_fingerprint": "...",
        "oauth_client": "ssh-proxy"
      }
    },
```

The `v2/info` endpoint belongs to the deprecated CF API v2. It shall not be used anymore and new features won't get implemented in v2.
CLI v6 is based on CF API v2 (and uses the `v2/info` endpoint) but CLI v6 is out of support: https://github.com/cloudfoundry/cli/wiki/Versioning-and-Support-Policy

I suggest to base this RFC on CF API v3 and CLI v8 only.

Reply:
hi @stephanme thanks for your input. I think I've misread the code looking for where the CLI gets the SSH endpoint from. There is a [getSSHEndpointInfo](https://github.com/cloudfoundry/cli/blob/main/cf/commands/application/ssh.go#L178) function that's still in the main branch of the CLI. However, I've tested the v8 CLI with `-v` and it does indeed call the `/` endpoint of CC.

I will update the RFC accordingly, thanks!

Reply:
Pushed a commit to change the compatibility level to v3 only.